components/brave_sync/client/bookmark_change_processor_unittest.cc

```diff
@@ -1466,3 +1467,110 @@ TEST_F(BraveBookmarkChangeProcessorTest, SyncTimestampMetaUpdateWay) {
   EXPECT_EQ(node_b_sync_timestamp,
       std::to_string(timestamp_resolve.ToJsTime()));
 }
+
+namespace {
+
+size_t g_overridden_minutes = 0;
+std::unique_ptr<base::subtle::ScopedTimeClockOverrides> OverrideForMinutes(
+    int overridden_minutes) {
+  g_overridden_minutes = overridden_minutes;
+  return std::make_unique<base::subtle::ScopedTimeClockOverrides>(
+    []() {
+     return base::subtle::TimeNowIgnoringOverride() +
+         base::TimeDelta::FromMinutes(g_overridden_minutes);
+    },
+    nullptr,
+    nullptr);
+}
+
+}  // namespace
+
+TEST_F(BraveBookmarkChangeProcessorTest, ExponentialResend) {
+  change_processor()->Start();
+  bookmarks::AddIfNotBookmarked(model(),
+                                 GURL("https://a.com/"),
+                                 base::ASCIIToUTF16("A.com"));
+
+  std::string send_retry_number;
+  GetSingleNodeByUrl(model(), "https://a.com/")->GetMetaInfo(
+     "send_retry_number", &send_retry_number);
+  EXPECT_TRUE(send_retry_number.empty());
+
+  int expected_send_retry_number = 0;
+
+  std::string record_a_object_id;
+
+  static const std::vector<int> exponential_waits = {0, 10, 20, 40, 80};
+  const int max_send_retries = exponential_waits.size() - 1;
+  std::set<int> should_sent_at;
+  size_t current_sum = 0;
+  for (size_t j = 0; j < exponential_waits.size(); ++j) {
+    current_sum += exponential_waits[j];
+    should_sent_at.insert(current_sum);
+  }
+  auto contains = [](const std::set<int>& set, int val) {
+    return set.find(val) != set.end();
+  };
+  EXPECT_TRUE(contains(should_sent_at, 0));
+  EXPECT_TRUE(contains(should_sent_at, 10));
+  EXPECT_TRUE(contains(should_sent_at, 30));
+  EXPECT_TRUE(contains(should_sent_at, 70));
+  EXPECT_TRUE(contains(should_sent_at, 150));
+  size_t i = 0;
+  for (; i <= 151; ++i) {
+    auto time_override = OverrideForMinutes(i);
+    bool is_send_expected = contains(should_sent_at, i);
+    int expect_call_times = is_send_expected ? 1 : 0;
+    EXPECT_CALL(*sync_client(), SendSyncRecords("BOOKMARKS", _)).Times(
+        expect_call_times);
+    EXPECT_CALL(*sync_client(), ClearOrderMap()).Times(1);
```

Comment:
this seems like leftover. And we can't clear order for every send or some records will miss their parent order

Reply:
Yes, this is leftover since https://github.com/brave/brave-core/blob/master/components/brave_sync/client/bookmark_change_processor.cc#L964 , https://github.com/brave/brave-core/pull/1174 .

Tried to modify it with https://github.com/brave/brave-core/pull/2986#discussion-diff-307946465R1020 , but had doubts.

So the correct place of `sync_client_->ClearOrderMap();` is as it is just in master, in the very end of `BookmarkChangeProcessor::SendUnsynced`
The description of the problem ClearOrderMap() solves: 
background.js should keep orders map while processing all of the 1000 bookmark records  chunks and clear the map after the last of chunks is sent.

So in the test-case above should not be invoked, but this require to fix `BookmarkChangeProcessor::SendUnsynced` and invoke `ClearOrderMap()` only if `SendSyncRecords` was called at least once, but do not call  `ClearOrderMap()` if `SendSyncRecords` had not been called.

Doing fix.

Reply:
Should be fixed with https://github.com/brave/brave-core/pull/2986/commits/b0c61baf0d99b3162a3c3601168b6e02397da858 and https://github.com/brave/brave-core/pull/2986/commits/55d4fb10bc01dd19ebb6fe4c9fdc8584758b6e86 .
Thanks @darkdh